### PR TITLE
[15.0][ADD] mail_no_user_assign_notification: New module

### PR DESCRIPTION
**Pull request by victoralmau**
New module

Please @pedrobaeza and @carolinafernandez-tecnativa can you review it?

@Tecnativa TT49257

### CODE REVIEW

File: mail_no_user_assign_notification/readme/USAGE.rst [hunk 1]
```diff
@@ -0,0 +1,4 @@
+#. Go to Purchase.
```
Comment (carolinafernandez-tecnativa) [inline mail_no_user_assign_notification/readme/USAGE.rst]: Just a suggestion,  put this as an example, as it is generic for all modules that has user_id.

**Comment by carolinafernandez-tecnativa** [comment]
The only user notification that is not working is when assigned new tasks as it is a many2many field and you can assign multiple users. 

**Configuration:**
image

**Mail received:**
image

I have checked with purchase order and sales order and leads works perfectly!

**Comment by victoralmau** [comment]
Tasks are out of the scope of this module; moreover, it is not just adding the `user_ids` field because the `project` module uses another “_different_” mechanism.

**Review by carolinafernandez-tecnativa — APPROVED** [review]

**Comment by HaraldPanten** [comment]
Hi @victoralmau @pedrobaeza I think we did something quite similar (we did it for V16) Maybe you want to have a look and both try to get the best from each module? Maybe we could join efforts --> https://github.com/sygel-technology/sy-social/tree/16.0/mail_block_user_assigned_message

File: mail_no_user_assign_notification/readme/CONFIGURE.rst [hunk 2]
```diff
@@ -0,0 +1,5 @@
+#. Go to Settings > Technical > System Parameters
+#. Edit "mail_no_user_assign_notification.models" key record
+#. Set any mode* (for example: res.partner).
```
Comment (chienandalu) [inline mail_no_user_assign_notification/readme/CONFIGURE.rst]:
```suggestion
#. Set any model (for example: res.partner).
```

**Comment by victoralmau** [comment]
> Hi @victoralmau @pedrobaeza I think we did something quite similar (we did it for V16) Maybe you want to have a look and both try to get the best from each module? Maybe we could join efforts --> https://github.com/sygel-technology/sy-social/tree/16.0/mail_block_user_assigned_message

I have reviewed what you indicate but I see that (although similar) they are different approaches (apply to the user_id field of the models defined in a system parameter, or indicate for each partner the model on which you want to do it), so I think they should be different modules.

File: mail_no_user_assign_notification/readme/DESCRIPTION.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+This module inhibits the auto-assign notification from the models that we want.
```
Comment (pedrobaeza) [inline mail_no_user_assign_notification/readme/DESCRIPTION.rst]:
```suggestion
This module inhibits the assignation notification from the models that we want.
```

File: mail_no_user_assign_notification/readme/CONFIGURE.rst [hunk 4]
```diff
@@ -0,0 +1,5 @@
+#. Go to Settings > Technical > System Parameters
+#. Edit "mail_no_user_assign_notification.models" key record
+#. Set any model* (for example: res.partner).
+
+It is possible to set several models separated by comma (e.g.: model1,model2).
```
Comment (pedrobaeza) [inline mail_no_user_assign_notification/readme/CONFIGURE.rst]:
```suggestion
NOTE1: It is possible to set several models separated by comma (e.g.: model1,model2).

NOTE2: The model should have the field `user_id` for this to be effective.
```

File: mail_no_user_assign_notification/readme/USAGE.rst [hunk 5]
```diff
@@ -0,0 +1,4 @@
+#. Go to Contacts.
+#. Create a partner and set another user as Salesperson.
+#. User will not receive the automatic assignment email.
+#. The same applies if another partner is modified.
```
Comment (pedrobaeza) [inline mail_no_user_assign_notification/readme/USAGE.rst]: Remember to not put a specific example, but the generic steps:

```suggestion
#. Go to one of the configured models (like Contacts).
#. Create or edit a record, and assign the field for the user (in contacts, it's called "Salesperson", in purchase orders it's "Purchase representative", etc).
#. The user won't receive the automatic assignment email.
```

File: mail_no_user_assign_notification/tests/test_mail_no_user_assign_notification.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+    def test_partner_create(self):
+        partner = self.env["res.partner"].create(
+            {
+                "name": "Test partner",
+                "user_id": self.user.id,
+            }
+        )
+        self.assertNotIn(
+            self.user.partner_id, partner.message_follower_ids.mapped("partner_id")
```
Comment (pedrobaeza) [inline mail_no_user_assign_notification/tests/test_mail_no_user_assign_notification.py]: This is a side effect of what we are doing, but you shouldn't test this.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

**Comment by pedrobaeza** [comment]
Add `ROADMAP.rst` for commenting that this is not valid for project tasks users assignation, as the field is multi-relational. A specific extension would be needed.

**Comment by victoralmau** [comment]
Changes done.

File: mail_no_user_assign_notification/readme/CONFIGURE.rst (hunk 4 above)
Comment (pedrobaeza) [reply mail_no_user_assign_notification/readme/CONFIGURE.rst]: Duplicated text

File: mail_no_user_assign_notification/readme/ROADMAP.rst [hunk 7]
```diff
@@ -0,0 +1,2 @@
+This is not valid for project tasks users assignation, as the field is multi-relational.
+A specific extension would be needed.
```
Comment (pedrobaeza) [inline mail_no_user_assign_notification/readme/ROADMAP.rst]:
```suggestion
* This is not valid for project tasks users assignation, as the field is multi-relational.
  A specific extension would be needed.
```

File: mail_no_user_assign_notification/tests/test_mail_no_user_assign_notification.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+        )
+        cls.icp.value = "res.partner"
+
+    def test_purchase_order_write(self):
+        partner = self.env["res.partner"].create({"name": "Test partner"})
+        self.assertNotIn(
+            self.user.partner_id, partner.message_follower_ids.mapped("partner_id")
+        )
+        partner.write({"user_id": self.user.id})
+        self.assertNotIn(
```
Comment (pedrobaeza) [inline mail_no_user_assign_notification/tests/test_mail_no_user_assign_notification.py]: This still testing something different from the purpose of the module.

File: mail_no_user_assign_notification/models/mail_thread.py [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    def _message_auto_subscribe_followers(self, updated_values, default_subtype_ids):
+        """Remove user_id from updated_values if this model has been set.
+        This will prevent the user from auto subscrube to records on creation and/or
+        assignment."""
+        if updated_values.get("user_id"):
+            icp = self.sudo().env["ir.config_parameter"]
+            models_to_skip = icp.get_param(
+                "mail_no_user_assign_notification.models", ""
+            )
+            models_to_skip = models_to_skip.split(",")
```
Comment (pedrobaeza) [inline mail_no_user_assign_notification/models/mail_thread.py]:
```suggestion
            models_to_skip = [x.strip() for x in models_to_skip.split(",")]
```

File: mail_no_user_assign_notification/tests/test_mail_no_user_assign_notification.py (hunk 8 above)
Comment (victoralmau) [reply mail_no_user_assign_notification/tests/test_mail_no_user_assign_notification.py]: Ok, changed the test approach.

**Review by pedrobaeza — APPROVED** [review]
/ocabot merge nobump

**Pull request merged**